sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala

```diff
@@ -437,16 +437,20 @@ object VirtualColumn {
  * The internal representation of the hidden metadata struct:
  * set `__metadata_col` to `true` in AttributeReference metadata
  * - apply() will create a metadata attribute reference
- * - unapply() will check if an attribute reference is the metadata attribute reference
+ * - unapply() will check if an attribute reference is the file source metadata attribute reference
  */
 object MetadataAttribute {
```

Comment:
Maybe we should have two objects
1. `MetadataAttribute`: This is used to create and extract general metadata attributes (not only for file source)
2. `FileSourceMetadataAttribute`: To extract file source metadata attributes

Reply:
Or we make sure that we only call `MetadataAttribute.unapply` in places that only deal with file sources.

Reply:
make sense!